[Spark] DV Reads Performance Improvement in Delta by removing Broadcasting DV Information

longvu-db opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-io/delta/blob/master/CONTRIBUTING.md
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
  6. If applicable, include the corresponding issue number in the PR title and link it in the body.
-->

#### Which Delta project/connector is this regarding?
<!--
Please add the component selected below to the beginning of the pull request title
For example: [Spark] Title of my pull request
-->

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

<!--
- Describe what this PR changes.
- Describe why we need the change.
 
If this PR resolves an issue be sure to include "Resolves #XXX" to correctly link and close the issue upon merge.
-->
Back then, we relied on an [expensive Broadcast of DV files](https://github.com/delta-io/delta/pull/1542) to pass the DV files to the associated Parquet Files. With the introduction of [adding custom metadata to files](https://github.com/apache/spark/pull/40677) introduced in Spark 3.5, we can now pass the DV through the custom metadata field, this is expected to improve the performance of DV reads in Delta.
## How was this patch tested?

<!--
If tests were added, say they were added here. Please make sure to test the changes thoroughly including negative and positive cases if possible.
If the changes were tested in any way other than unit tests, please clarify how you tested step by step (ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future).
If the changes were not tested, please explain why.
-->
Adjusted the existing UTs that cover our changes.
## Does this PR introduce _any_ user-facing changes?
No.
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Delta Lake versions or within the unreleased branches such as master.
If no, write 'No'.
-->

xupefei commented on spark/src/main/scala/org/apache/spark/sql/delta/files/TahoeFileIndex.scala:
```diff
@@ -126,6 +127,19 @@ abstract class TahoeFileIndex(
+      metadata.put(DeltaParquetFileFormat.FILE_ROW_INDEX_FILTER_ID_ENCODED,
+        JsonUtils.toJson(addFile.deletionVector))
+    }
+    val filterType = rowIndexFilters.getOrElse(Map.empty)
+      .getOrElse(addFile.path, RowIndexFilterType.IF_CONTAINED)
+    metadata.put(DeltaParquetFileFormat.FILE_ROW_INDEX_FILTER_TYPE, filterType)
```
This will add one metadata entry regardless the existence of DVs on thi file, right?
How about not adding it but later we do `partitionedFile.otherConstantMetadataColumnValues.contains(FILE_ROW_INDEX_FILTER_TYPE)`?

longvu-db replied:
@xupefei True that

longvu-db replied:
Although the map.get and isDefined also do the `contains` trick

xupefei commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala:
```diff
@@ -33,26 +29,31 @@ import org.apache.parquet.hadoop.ParquetFileReader
 
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    enableDeletionVectorsInNewTables(spark.conf)
```
Why enable DV for all tests?

longvu-db replied:
@xupefei What do you mean sorry? Could you further elaborate?

longvu-db replied:
@xupefei I added some tests with DV turned off, would you be so kind to take another look at the PR :)?

xupefei commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala:
```diff
@@ -112,11 +109,7 @@ class DeltaParquetFileFormatSuite extends QueryTest
-            case RowIndexFilterType.IF_NOT_CONTAINED => (0, 1)
-            case _ => (-1, -1) // Invalid, expecting the test to fail.
-          }
+          val (deletedColumnValue, notDeletedColumnValue) = (1, 0)
```
You reduced the coverage because it's partially covered by CDCReader tests, isn't it?
Anyway have more tests are always better. I would prefer to keep it.

longvu-db replied:
@xupefei Hmm, but right now I don't have control over the rowIndexFilterType though.

Which case does the CDCReader not cover?

xupefei replied:
Ahha I see. We no longer broadcast DVs so there's no IF_CONTAINED or IF_NOT_CONTAINED at this level.

xupefei commented on spark/src/main/scala/org/apache/spark/sql/delta/DeltaParquetFileFormat.scala:
```diff
@@ -281,27 +295,28 @@ case class DeltaParquetFileFormat(
+      val filterTypeOpt = partitionedFile.otherConstantMetadataColumnValues
+        .get(FILE_ROW_INDEX_FILTER_TYPE)
+      if (dvDescriptorOpt.isDefined && filterTypeOpt.isDefined) {
+        val rowIndexFilter = if (filterTypeOpt.get == RowIndexFilterType.IF_CONTAINED) {
+          DropMarkedRowsFilter
+        } else if (filterTypeOpt.get == RowIndexFilterType.IF_NOT_CONTAINED) {
+          KeepMarkedRowsFilter
```
Case match is better, because we won't call Option.get multiple times.

xupefei reviewed (commented):
Very nice, thank you! Looking forward to seeing how much speed-up we can get with this change :)

longvu-db commented on spark/src/main/scala/org/apache/spark/sql/delta/DeltaParquetFileFormat.scala:
```diff
@@ -283,27 +297,31 @@ case class DeltaParquetFileFormat(
+      } else if (dvDescriptorOpt.isDefined || filterTypeOpt.isDefined) {
+        throw new IllegalStateException(
+          s"Both ${FILE_ROW_INDEX_FILTER_ID_ENCODED} and ${FILE_ROW_INDEX_FILTER_TYPE} " +
+            "should both have a value or not have a value together")
```
Add a period here @longvu-db

xupefei commented on spark/src/main/scala/org/apache/spark/sql/delta/DeltaParquetFileFormat.scala:
```diff
@@ -205,16 +203,27 @@ case class DeltaParquetFileFormat(
+        super.metadataSchemaFields ++ rowTrackingFields
+      }
+
+    val isDVsEnabled = DeltaConfigs.ENABLE_DELETION_VECTORS_CREATION.fromMetaData(metadata)
```
`metadata` shows the current status of the table, i.e., if DV is enabled in the current version, isn't it? How about some old versions that have DV, but now the user has decided to disable it? I think we should look at the table protocol: if DV table feature is turned on, then we assume the table contains DV.

longvu-db replied:
@xupefei Quick question, can an existing table add new table feature? Like if a table does not have DV from the beginning, in the property, and then at some point later in time get added DVs.

longvu-db replied:
@xupefei I think you can do this using ALTER TABLE right? But once it is in, you cannot drop it at the moment.

xupefei reviewed (commented):
After looking at the code, I feel that we need a test case with the following logic:

1. Create a table without DV. Insert some values.
2. Turn on DV (`delta.enableDeletionVectors` = `true`).
3. Delete some values.
4. Turn off DV (`delta.enableDeletionVectors` = `false`).
5. Delete some values.

Check:
1. Version 1 does not contain DV metadata column.
2. Version 2 does not contain DV metadata column (because there's no DV in this table - correct me if I'm wrong).
3. Version 3 does contain DV metadata columns.
3. Version 4 does contain DV metadata columns.
3. Version 5 does contain DV metadata columns.

longvu-db commented:
@xupefei I addressed your comments =)).

Regarding the new test case:
1. Version 2 **should contain** DV metadata column (because having the metadata columns depends on whether or not the table's protocol supports DV, not on whether any DVs is written or not)
2. What is the value of doing action 5? If we disable the DV for the table metadata, and if the DV metadata indeed stay, then whatever actions we do, it would always stay right?

xupefei commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala:
```diff
@@ -16,140 +16,214 @@
+
+  private val FILE_ROW_INDEX_FILTER_TYPE = "_metadata.row_index_filter_type"
+
+  private def checkDVMetadataFields(tablePath: String, checkMetadataFields: Boolean): Unit = {
```
```suggestion
  private def checkDVMetadataFields(tablePath: String, metadataContainsDVFields: Boolean): Unit = {
```

johanl-db commented on spark/src/main/scala/org/apache/spark/sql/delta/DeltaParquetFileFormat.scala:
```diff
@@ -104,7 +102,7 @@ case class DeltaParquetFileFormat(
     sparkSession: SparkSession, options: Map[String, String], path: Path): Boolean = isSplittable
 
-  def hasDeletionVectorMap: Boolean = broadcastDvMap.isDefined && broadcastHadoopConf.isDefined
+  def hasBroadcastHadoopConf: Boolean = broadcastHadoopConf.isDefined
```
I think you can get rid of `broadcastHadoopConf` and `hasBroadcastHadoopConf` altogether here.

The only use we have for it now is to pass it to `rowIndexFilter.createInstance` when creating the DV filter but we also have `hadoopConf: Configuration` around (passed as an argument to `buildReaderWithPartitionValues`) which I believe should be sufficient here: we don't need this configuration to be broadcast first for the purpose of loading DVs from storage

longvu-db replied:
@johanl-db No clue why we had it back in the days hmm

longvu-db replied:
@johanl-db Removing `broadcastHadoopConf` doesn't seem to work, I found this [error message](https://runbot-ci.cloud.databricks.com/build/Delta-OSS-Scala212-Pr/run/17985292)but I haven't investigated yet, changing back to using the `broadcastHadoopConf` makes the tests passed

xupefei commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala:
```diff
@@ -16,140 +16,214 @@
+      // Version 1: Create a table without DV and insert some values.
+      generateData(tablePath)
+
+      checkDVMetadataFields(tablePath, checkMetadataFields = false)
```
Can we also check the table contents using `checkResult` each time after we check the metadata?

xupefei reviewed (commented):
A few improvements to the tests then we're good to go

xupefei commented:
Cc @larsk-db

johanl-db commented on spark/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala:
```diff
@@ -93,20 +91,22 @@ object ScanWithDeletionVectors {
 
     // See if the relation is already modified to include DV reads as part of
     // a previous invocation of this rule on this table
-    if (fileFormat.hasDeletionVectorMap) return None
+    if (fileFormat.hasBroadcastHadoopConf) return None
```
We should update that check since using a broadcast conf isn't a marker of the preprocess step for DVs (we actually remove this need in this change..)

We could instead check if the scan already contains the `IS_ROW_DELETED_STRUCT_FIELD` field, e.g.:


```
    if (scan.output.exists(_.name == IS_ROW_DELETED_COLUMN_NAME)) return None
```

johanl-db commented on spark/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala:
```diff
@@ -145,7 +144,7 @@ object ScanWithDeletionVectors {
       new SerializableConfiguration(tahoeFileIndex.deltaLog.newDeltaHadoopConf()))
 
     val newFileFormat = fileFormat.copyWithDVInfo(
-      tahoeFileIndex.path.toString, filePathToDVBroadcastMap, hadoopConfBroadcast)
+      tahoeFileIndex.path.toString, hadoopConfBroadcast)
```
I would remove `copyWithDVInfo` at this point and directly do here:
```
val newFileFormat = fileFormat.copy(isSplittable = false, disablePushDowns = true)
```


Or rename `copyWithDVInfo` to e.g. `disableSplittingAndPushdown`

johanl-db commented on spark/src/main/scala/org/apache/spark/sql/delta/files/TahoeFileIndex.scala:
```diff
@@ -126,6 +127,19 @@ abstract class TahoeFileIndex(
     addFile.defaultRowCommitVersion.foreach(defaultRowCommitVersion =>
       metadata.put(DefaultRowCommitVersion.METADATA_STRUCT_FIELD_NAME, defaultRowCommitVersion))
 
+    // Attach filter types to FileActions, so that later [[DeltaParquetFileFormat]] could pick it up
+    // to decide which kind of rows should be filtered out. This info is necessary for reading CDC
+    // rows that have been deleted (marked in DV), in which case marked rows must be kept rather
+    // than filtered out. In such a case, the `filterTypes` map will be populated by [[CDCReader]]
+    // to indicate IF_NOT_CONTAINED filter should be used. In other cases, `filterTypes` will be
+    // empty, so we generate IF_CONTAINED as the default DV behavior
```
Move this comment just before `val filterType`

Also it could be a bit shorter, e.g.:
> Set the filter type to `IF_CONTAINED` by default to filter out rows unless a filter type was explicitly provided in `rowIndexFilters`. This can happen e.g. when reading CDC data to keep deleted rows instead of filtering them out.

johanl-db replied:
Not addressed ^

longvu-db replied:
@johanl-db I actually addressed, but I was not sure if your version removes too much original info, @xupefei, is it fine if we keep it like

> Set the filter type to IF_CONTAINED by default to filter out rows unless a filter type was explicitly provided in rowIndexFilters. This can happen e.g. when reading CDC data to keep deleted rows instead of filtering them out.

xupefei replied:
I would suggest
> Set the filter type to IF_CONTAINED by default to let [[DeltaParquetFileFormat]] filter out rows unless a filter type was explicitly provided in rowIndexFilters. This can happen e.g. when reading CDC data to keep deleted rows instead of filtering them out.

johanl-db commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala:
```diff
@@ -16,140 +16,214 @@
     }
   }
 
+  testWithBothParquetReaders("Existence of Deletion Vectors metadata columns when we " +
```
This test probably becomes moot if you don't pass the DV information via the metadata struct anymore 9see my previous comment)

larsk-db commented:
> Cc @larsk-db

I took a quick look, but it seems you and @johanl-db are on top of this. Ping me if you have something specific that needs my input.

johanl-db commented on spark/src/main/scala/org/apache/spark/sql/delta/DeltaParquetFileFormat.scala:
```diff
@@ -417,6 +416,12 @@ object DeltaParquetFileFormat {
   val ROW_INDEX_COLUMN_NAME = "__delta_internal_row_index"
   val ROW_INDEX_STRUCT_FIELD = StructField(ROW_INDEX_COLUMN_NAME, LongType)
 
+  /** The name of the metadata column that contains the encoded row index filter identifier. */
```
Update this comment and the one below, this isn't the name of a metadata column anymore, just the key for the row index filter id / type in the file's `otherConstantMetadataColumnValues` map.

longvu-db replied:
@johanl-db My bad yeah

tdas reviewed (approved):
I highly approve this change :D
Thank you for doing this! This will improve the stability on large tables with lots of DVs significantly.
